### PR TITLE
Add aggregate targets initial support

### PR DESCRIPTION
**Pull request by MontakOleg**
Resolves https://github.com/tuist/tuist/issues/3388

### Short description 📝

Initial support for aggregate targets added. At the moment only scripts (build phases) is supported – this is usually what people need to run swiftlint / periphery / etc.

```
let project = Project(
    name: "App",
    targets: [
        Target(
            name: "App",
            platform: .macOS,
            product: .aggregateTarget,
            bundleId: "aggregate.bundle.id",
            scripts: [
                .pre(
                    path: "./Scripts/run_periphery_via_mint.sh",
                    name: "Periphery scan",
                    basedOnDependencyAnalysis: false
                )
            ]
        ),
    ]
)
```

You can find alternative approach (with aggregate target as separate model; not finished) in https://github.com/MontakOleg/tuist/tree/aggregate-targets-as-separate-model.

### How to test the changes locally 🧐

Run `test_generateAggregateTarget_actions` unit test.
Run `./fourier test tuist acceptance projects/tuist/features/generate-4.feature:21` acceptance test.

Explore / test `ios_app_with_aggregate_target` fixture manually.

### Contributor checklist ✅

- [x] The code has been linted using run `./fourier lint tuist --fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

**Comment by MontakOleg** [comment]
I am also thinking about syntax for end users. Now we use syntax for regular targets which requires to specify platform and bundleId.

We can create separate `AggregateTarget` model for manifest only with supported subset of properties for aggregate targets and then map it to regular Target(product: .aggregateTarget) on graph level. Then syntax could be something like:

```
let project = Project(
    name: "App",
    targets: [...],
    aggregateTargets: [
        AggregateTarget(
            name: "App",
            scripts: [
                ...
            ]
        )
    ]
)
```

@fortmarek, @pepicrft please let me know what do you think?

**Comment by pepicrft** [comment]
Hey @MontakOleg.
Thanks for going ahead with addressing this problem. 

I created an `Aggregate` target in Xcode and noticed it only has a very subset of the functionality a normal target has, and that makes me strongly inclined towards having its own type in `ProjectDescription` as you recommend. I'd align the naming and structure of the elements that overlap between `Target` and `Aggregate`.

Moreover, note that they support build settings so we need to expose an API for that too.

One thing that came to my mind, but that we don't need to address as part of this PR, is that we should extend the scripts API for developers to declare input and output files and environment variables that scripts depend on. This is critical for caching to take those scripts into account when calculating the fingerprint.

Please, let me know if that helps continuing with the work here or if there are still some open questions.

**Comment by pepicrft** [comment]
Side note: I made you contributor of the organization so that you don't have to work using forks, which is terribly inconvenient.

**Comment by MontakOleg** [comment]
Thank you for quick response! I will look how it will work with separate AggregateTarget model on manifest level.

**Comment by MontakOleg** [comment]
Added `AggregateTarget` model to the manifest. I intentionally restricted their properties to `name`, `platform`, `scripts` and `settings`.

Except properties above, looks like aggregate targets in Xcode also supports `buildRules` and `dependencies`:
- for `buildRules` I don't know good use case for them;
- for `dependencies` I don't want to deal with them at the moment (things become complicated quickly and I personally don't need them).

The main use case for aggregate targets for me: run some scripts (`swiftlint` / `periphery`) and see their output (warnings and errors) right in Xcode.

File: projects/tuist/fixtures/ios_app_with_aggregate_target/App/Info.plist
Comment (fortmarek) [inline projects/tuist/fixtures/ios_app_with_aggregate_target/App/Info.plist:1]: How have you created this project? Shouldn't the Info.plist be optional nowadays? I think we should consider not adding it to new fixtures (and ideally remove it from old ones).

File: projects/tuist/fixtures/ios_app_with_aggregate_target/.gitignore
Comment (fortmarek) [inline projects/tuist/fixtures/ios_app_with_aggregate_target/.gitignore:1]: We should be able to add a `.gitignore` at the `fixtures` level that will do this matching for any fixture in the directory, so we don't have to add this for every new fixture.

**Review by fortmarek — APPROVED** [review]
I'm aligned with the API and the code looks great, thanks for the contribution 👏 Just have two comments around the fixture.

File: Sources/ProjectDescription/AggregateTarget.swift [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import Foundation
+
+/// An aggregate target of a project.
+public struct AggregateTarget: Codable, Equatable {
```
Comment (danyf90) [inline Sources/ProjectDescription/AggregateTarget.swift:4]: Why do we need a new type, can't it just be a `product`?
Wouldn't it be more consistent with what SPM does?

**Review by danyf90 — CHANGES_REQUESTED** [review]

File: Sources/ProjectDescription/AggregateTarget.swift (hunk 1 above)
Comment (fortmarek) [reply Sources/ProjectDescription/AggregateTarget.swift:4]: See the reasoning above: https://github.com/tuist/tuist/pull/5266#issuecomment-1614934129

Alternatively, we need to change the `Target` itself. We can make bundle id and platform properties optional and provide two different inits or builders – one for aggregate and one for non-aggreagate targets (where we'd require the bundle id and platform, so users can't make invalid targets).

Comment (danyf90) [reply Sources/ProjectDescription/AggregateTarget.swift:4]: Makes sense 👍

File: projects/tuist/fixtures/ios_app_with_aggregate_target/.gitignore
Comment (MontakOleg) [reply projects/tuist/fixtures/ios_app_with_aggregate_target/.gitignore:1]: We have (pretty simple) `.gitignore` on fixtures level and it works just fine for my fixture 🙂:

```
*.xcodeproj
*.xcworkspace
```

So I deleted `.gitignore` and other unneeded files from new fixture.

File: projects/tuist/fixtures/ios_app_with_aggregate_target/App/Info.plist
Comment (MontakOleg) [reply projects/tuist/fixtures/ios_app_with_aggregate_target/App/Info.plist:1]: Done

Comment (kwridan) [reply projects/tuist/fixtures/ios_app_with_aggregate_target/App/Info.plist:1]: Note: for the older fixtures, it's still worth keeping a handful with explicit `Info.plist` files to verify specifying files still works (it's a supported pattern in addition to the auto-generated one by Tuist).

**Review by kwridan — COMMENTED** [review]
Thanks @MontakOleg 

For your aggregate target use case, do you explicitly select its scheme and run that or have a target dependency which indirectly runs it?

For the latter, I suspect some form of dependency support may be needed (may complicate the integration slightly - so do agree of keeping this simple for the time being without it)

File: Sources/TuistCore/Graph/ModelExtensions/Product+Core.swift [hunk 2]
```diff
@@ -51,6 +51,8 @@ extension Product {
             return .systemExtension
         case .extensionKitExtension:
             return .extensionKitExtension
+        case .aggregateTarget:
+            return .none
```
[the remark this answers is not in the document]
Comment (MontakOleg) [reply Sources/TuistCore/Graph/ModelExtensions/Product+Core.swift:55]: @fortmarek [suggested](https://github.com/tuist/tuist/issues/3388#issuecomment-1129645454) trying 2 approaches to add aggregate targets:
1. By reusing existing Target type with new product type = .aggregateTarget
2. By creating new separate type AggregateTarget

I started with second approach, it became a little bit complex and felt fragile after adding `AggregateTarget` into `Graph`. I had basic implementation, but it still lacked some of the features, for example autogenerated schemes. You can found this attempt here – https://github.com/MontakOleg/tuist/tree/aggregate-targets-as-separate-model

With first approach it much simpler. At the same time we have created separate interface with AggregateTarget for end users, so we can change internal implementation later if we decide that it works better.

**Comment by MontakOleg** [comment]
Hi @kwridan!

>For your aggregate target use case, do you explicitly select its scheme and run that or have a target dependency which indirectly runs it?

We usually run them manually.

For us it's 2 use cases :

1. Move out linters / code generators / etc. from main target build phases to speedup incremental builds (inspired by https://www.merowing.info/improve-build-times-by-extracting-3rd-party-tooling-to-processing-script/). Instead run them in git-hook / on CI / or manually from Xcode schema.

2. Run slow scripts like Periphery and see their output in Xcode

>For the latter, I suspect some form of dependency support may be needed (may complicate the integration slightly - so do agree of keeping this simple for the time being without it)

I would prefer to keep it simple and not introduce dependencies for now.

P.S. If you really need dependencies, as a temporary workaround, you can use regular .framework target with no sources instead of aggregate target. It should work fine.

File: projects/tuist/fixtures/ios_app_with_aggregate_target/.gitignore
Comment (fortmarek) [reply projects/tuist/fixtures/ios_app_with_aggregate_target/.gitignore:1]: Thank you ❤️

**Review by danyf90 — APPROVED** [review]

**Review by kwridan — COMMENTED** [review]
Thanks for the explanation @MontakOleg 

> @fortmarek https://github.com/tuist/tuist/issues/3388#issuecomment-1129645454 trying 2 approaches to add aggregate targets:
> 
> By reusing existing Target type with new product type = .aggregateTarget
> By creating new separate type AggregateTarget
>
> I started with second approach, it became a little bit complex and felt fragile after adding AggregateTarget into Graph. I had basic implementation, but it still lacked some of the features, for example autogenerated schemes. You can found this attempt here – https://github.com/MontakOleg/tuist/tree/aggregate-targets-as-separate-model
> 
> With first approach it much simpler. At the same time we have created separate interface with AggregateTarget for end users, so we can change internal implementation later if we decide that it works better.

+1 for keeping the public API the desired one as it does give some options - the bit I'm a bit worried about is, the aggregate targets are now indirectly part of the `Graph` as `macOS` targets with an unsupported `productType` which we'll need to always account for / work around going forward internally 😞 

I had a look at your branch where the other approach was attempted and it looked good 👍 from what I saw the areas that needed some extra work were:

- Omitting the aggregate targets from cache / tree shaker logic (don't think they are needed there - and may fix some of the issues where the generated project is empty)
- Autogenerated schemes mapper to include aggregate targets
- Scheme generator to accommodate aggregate targets 

Were there any other concerns / issues you were concerned about?

File: Sources/TuistGenerator/Generator/SchemeDescriptorsGenerator.swift [hunk 3]
```diff
@@ -872,7 +875,7 @@ final class SchemeDescriptorsGenerator: SchemeDescriptorsGenerating {
     /// - Returns: Buildable reference.
     private func targetBuildableReference(
         target: Target,
-        pbxTarget: PBXNativeTarget,
+        pbxTarget: PBXNativeTarget?,
```
[the remark this answers is not in the document]
Comment (MontakOleg) [reply Sources/TuistGenerator/Generator/SchemeDescriptorsGenerator.swift]: Good catch! I compared project files before and after opening with `xcdiff`, but it looks like it does not show scheme changes.
Added `pbxTarget` for aggregate targets too now.

**Comment by MontakOleg** [comment]
>the bit I'm a bit worried about is, the aggregate targets are now indirectly part of the Graph as macOS targets with an unsupported productType which we'll need to always account for / work around going forward internally

Having `AggregateTarget` as separate type looks 'safer' – it's additive change, we don't break anything related to `Target`.
At the same time compiler don't enforce / advice us to process aggregate targets (alongside with regular targets) in different places in the project, so we can easily forget about them, and some `tuist` functionality will not work with them. That what I mean when say that it feels fragile.

>Omitting the aggregate targets from cache / tree shaker logic (don't think they are needed there - and may fix some of the issues where the generated project is empty)

I added aggregate targets to tree shaker logic because without it aggregate target was eliminated by some reason on my work (big) project. However it worked fine on simple fixture project. Maybe its better to investigate further if we want to proceed with separate `AggregateTarget` approach.

>Autogenerated schemes mapper to include aggregate targets

Have not looked at this code yet, but I think it should be pretty easy.

>Scheme generator to accommodate aggregate targets

As I remember it already worked, but I can double check.

>Were there any other concerns / issues you were concerned about?

One more concern was: what if we will want to add dependencies support between native and aggregate targets, how many code will be affected? How not to miss anything without compiler support?

**Comment by MontakOleg** [comment]
As more I think about it, as less I want to add all this complexity into tuist 😀 We get not so much from having aggregate targets – they can be replaced/worked around with regular targets.

**Comment by danyf90** [comment]
If we don't think the additional complexity is not worth, let's close this for now.
Thanks a lot for your effort on this @MontakOleg 🙏

**Pull request closed without merging**